include/revng/BasicAnalyses/GeneratedCodeBasicInfo.h

```diff
@@ -250,41 +243,77 @@ class GeneratedCodeBasicInfo : public llvm::ModulePass {
 
   /// \brief Return the program counter of the next (i.e., fallthrough)
   ///        instruction of \p TheInstruction
-  MetaAddress getNextPC(llvm::Instruction *TheInstruction) const {
+  MetaAddress getNextPC(llvm::Instruction *TheInstruction) {
     auto Pair = getPC(TheInstruction);
     return Pair.first + Pair.second;
   }
 
-  llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) const {
+  static llvm::CallInst *getFunctionCall(llvm::BasicBlock *BB) {
     return getFunctionCall(BB->getTerminator());
   }
 
   // TODO: is this a duplication of FunctionCallIdentification::isCall?
   // TODO: we could unpack the information too
-  llvm::CallInst *getFunctionCall(llvm::Instruction *T) const {
+  static llvm::CallInst *getFunctionCall(llvm::Instruction *T) {
+    revng_assert(T != nullptr);
     revng_assert(T->isTerminator());
-    auto It = T->getIterator();
-    auto End = T->getParent()->begin();
-    while (It != End) {
-      It--;
-      if (llvm::CallInst *Call = getCallTo(&*It, "function_call"))
+    llvm::Instruction *Previous = getPrevious(T);
+    while (Previous != nullptr && isMarker(Previous)) {
+      if (auto *Call = getCallTo(Previous, "function_call"))
         return Call;
 
-      if (not isMarker(&*It))
-        return nullptr;
+      Previous = getPrevious(Previous);
     }
 
     return nullptr;
   }
 
-  bool isFunctionCall(llvm::BasicBlock *BB) const {
+  static bool isFunctionCall(llvm::BasicBlock *BB) {
     return isFunctionCall(BB->getTerminator());
   }
 
-  bool isFunctionCall(llvm::Instruction *T) const {
+  static bool isFunctionCall(llvm::Instruction *T) {
     return getFunctionCall(T) != nullptr;
   }
 
+  static llvm::BasicBlock *getSuccessor(llvm::BasicBlock *BB) {
+    return getSuccessor(BB->getTerminator());
+  }
+
+  /// \brief Return the successor basic block given a function_call marker.
+  static llvm::BasicBlock *getSuccessor(llvm::Instruction *T) {
+    if (auto *Call = getFunctionCall(T)) {
+      if (auto *Successor = llvm::dyn_cast<llvm::BlockAddress>(Call->getOperand(0)))
+        return Successor->getBasicBlock();
+    }
+
+    return nullptr;
+  }
```

Comment:
"Successor" does not seem like an appropriate name. What about Callee?
Also, did you see `getFunctionCallCallee`?
There's quite some duplication between GCBI and `revng.h`/`IRHelpers.h`. Let's not make this worse.

Reply:
Callee sounds better, but I didn't use it because there's already `getCallee` in IRHelpers.h. 
`getFunctionCallCallee` does the same job but it seems to be only used within the StackAnalysis, and since GCBI.h already had `getFunctionCall` implemented, it looked much cleaner to me to have that `getSuccessor` method. Maybe we can drop `getFunctionCallCallee` from IRHelpers.h and move `getSuccessor` to `getCallee` (now that it is static)?

Reply:
Let's move as many `static` methods of GCBI into `IRHelpers.h`.
After all they're just helper functions.
One day we'll reorganize IRHelpers.h, but let's not make grow GCBI.

Reply:
Shouldn't it be better the opposite? `getFunctionCall` was already in GBCI.h, and given that `getFallthrough` is now there, and this point `getSuccessor` should be left there too, since both using `getFunctionCall`. This way everything is more concise and clean (and logically speaking would make sense to keep them there). Or do you want both of them to be in IRHelpers.h?

Reply:
Thins were in part in `IRHelpers.h` and in part in `GCBI`. Let's move stuff in `IRHelpers.h`.
Those functions are useful even if you're not in a pass. And if you're not in a pass there's no point in importing GCBI.h.
Anyway, this need further re-organization. But for now, move as much as you can in `IRHelpers.h`.

Reply:
Alright, sounds good, thanks for clarifying.